File: app/api/v1/booking_config.py
"""Booking schedule-grid config endpoints — two-level (platform + tenant) config.

Five endpoints, all under ``/bookings/config``:

  - GET  /platform              — super_admin only (platform-wide row or None)
  - PUT  /platform              — super_admin only (``settings:update`` is tenant-
                                  scoped, so the platform scope is gated on the
                                  stronger ``require_super_admin`` instead)
  - GET  /tenant/{tenant_id}    — the caller's own tenant row or None; super_admin
                                  can read any tenant; cross-tenant store roles → 403
  - PUT  /tenant/{tenant_id}    — upsert one tenant's override (``settings:update``);
                                  cross-tenant store roles → 403; super_admin → 200
  - GET  /effective?tenant_id=  — the resolved three-level fallback for the grid

Permission model (plan §4.3 + D5):

  - **Platform write**: super_admin only. ``settings:update`` is a *tenant*-
    scoped perm (seeded for owner/admin), so it cannot gate the platform scope —
    a tenant owner must NOT reach the platform-default write. Use
    ``require_super_admin`` for both platform GET and PUT (reading the platform
    default is itself a platform-wide action; a store role has no business
    seeing it directly — they see it via ``/effective``).
  - **Tenant write**: ``settings:update``. super_admin short-circuits via the
    casbin bypass; owner/admin of *this* tenant pass; member/customer → 403.
    Cross-tenant writes are blocked in the service body: a store role's
    ``user.tenant_id`` must equal the path ``tenant_id``, else 403.
  - **Effective read**: any role that can read the target tenant's bookings.
    Platform writers (super_admin / hq_staff) MUST carry ``tenant_id`` (the
    target store); store roles MUST NOT carry it (anti-forgery — resolves to
    their own tenant). Mirrors ``resolve_target_tenant`` from the devices/
    bookings write path but in query-param form.
"""

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import (
    CurrentUser,
    get_current_user,
    require_permission,
    require_super_admin,
)
from app.core.database import get_db
from app.schemas.booking_config import (
    BookingConfigRead,
    BookingConfigUpsert,
    EffectiveBookingConfig,
)
from app.services.booking_config_service import booking_config_service
from app.services.permission_service import is_platform_writer

router = APIRouter(prefix="/bookings/config", tags=["bookings"])


# --------------------------------------------------------------- platform scope


@router.get(
    "/platform",
    response_model=BookingConfigRead | None,
    dependencies=[Depends(require_super_admin())],
)
async def get_platform_config(
    db: AsyncSession = Depends(get_db),
) -> BookingConfigRead | None:
    """The platform-wide default config (super_admin only). None if unset."""
    return await booking_config_service.get_platform(db)


@router.put(
    "/platform",
    response_model=BookingConfigRead,
    dependencies=[Depends(require_super_admin())],
)
async def upsert_platform_config(
    payload: BookingConfigUpsert,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> BookingConfigRead:
    """Create or update the platform-wide default config (super_admin only)."""
    return await booking_config_service.upsert_platform(db, payload, actor_id=user.user_id)


# ----------------------------------------------------------------- tenant scope


def _ensure_tenant_access(user: CurrentUser, tenant_id: str) -> None:
    """Block cross-tenant access for store roles.

    A platform writer (super_admin / hq_staff) can target any tenant. A store
    role (owner / admin / member / customer) can only touch their own tenant —
    a mismatched path ``tenant_id`` is a forgery attempt → 403. Mirrors the
    ``resolve_target_tenant`` rule from the devices/bookings write path.
    """
    if is_platform_writer(user.platform_role):
        return
    if user.tenant_id != tenant_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="无权限:只能操作本租户的配置",
        )


@router.get(
    "/tenant/{tenant_id}",
    response_model=BookingConfigRead | None,
    dependencies=[Depends(require_permission("settings", "read"))],
)
async def get_tenant_config(
    tenant_id: str,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> BookingConfigRead | None:
    """One tenant's override row. None if the tenant uses the platform default.

    super_admin can read any tenant; store roles are pinned to their own.
    """
    _ensure_tenant_access(user, tenant_id)
    return await booking_config_service.get_tenant(db, tenant_id)


@router.put(
    "/tenant/{tenant_id}",
    response_model=BookingConfigRead,
    dependencies=[Depends(require_permission("settings", "update"))],
)
async def upsert_tenant_config(
    tenant_id: str,
    payload: BookingConfigUpsert,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> BookingConfigRead:
    """Create or update one tenant's override (upsert).

    ``settings:update`` is seeded for owner/admin; super_admin bypasses casbin.
    Cross-tenant store roles → 403.
    """
    _ensure_tenant_access(user, tenant_id)
    return await booking_config_service.upsert_tenant(
        db, tenant_id, payload, actor_id=user.user_id
    )


# --------------------------------------------------------------- effective read


@router.get(
    "/effective",
    response_model=EffectiveBookingConfig,
)
async def get_effective_config(
    tenant_id: str | None = Query(
        default=None,
        description="目标门店。平台角色必带;门店角色禁带(防伪造)。",
    ),
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> EffectiveBookingConfig:
    """The resolved three-level fallback config for the grid.

    - Platform writers (super_admin / hq_staff): MUST carry ``tenant_id`` (the
      target store). Missing → 403.
    - Store roles: MUST NOT carry ``tenant_id`` (anti-forgery). The effective
      tenant is always their own. Carrying it → 403.
    """
    if is_platform_writer(user.platform_role):
        if not tenant_id:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="平台角色查询 effective 配置必须指定目标门店(tenant_id)",
            )
        target = tenant_id
    else:
        if tenant_id is not None:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="门店角色不可指定目标租户(tenant_id)",
            )
        target = user.tenant_id
    return await booking_config_service.get_effective(db, target)

File: frontend/src/api/endpoints/booking-config.ts
/**
 * endpoints/booking-config — booking schedule-grid config (booking-schedule-grid 切片 01).
 *
 * Extracted from endpoints.ts monolith (plan-queries-endpoints-domain-split.md).
 * Locality move; see ./core.ts for the shared client.
 */
import {
  api,
} from "../client";
import type {
  BookingConfig,
  BookingConfigEffective,
  BookingConfigUpsert,
} from "../types";
// ---------- booking schedule-grid config (booking-schedule-grid 切片 01) ----------
//
// Two-level configuration for the schedule grid, mirroring the backend
// /bookings/config router (app/api/v1/booking_config.py):
// - GET  /bookings/config/platform         — super_admin read of the platform
//                                            default row (null body when none)
// - PUT  /bookings/config/platform         — super_admin upsert of that row
// - GET  /bookings/config/tenant/{id}      — settings:read view of one store's
//                                            override (own-tenant enforced
//                                            server-side; null body when none)
// - PUT  /bookings/config/tenant/{id}      — settings:update upsert of that
//                                            override
// - GET  /bookings/config/effective        — merged view the grid renders off
//                                            (tenant → platform → hardcoded
//                                            fallback); ``tenant_id`` query is
//                                            REQUIRED for platform roles and
//                                            FORBIDDEN for store roles (anti-
//                                            forgery, enforced server-side)
//
// GET platform / tenant return ``BookingConfig | null`` (200 + null body when
// the row doesn't exist — NOT 404). PUT is a full replace of the three upsert
// fields (the backend upserts: create-if-absent). ``window_*`` are "HH:MM"
// strings so they pass straight through ``<input type="time">`` values.

/** GET /bookings/config/platform — the platform-wide default row, or null
 * when none is seeded yet. super_admin only (route-level guard). */
export async function fetchPlatformBookingConfig(): Promise<BookingConfig | null> {
  const { data } = await api.get<BookingConfig | null>(
    "/bookings/config/platform",
  );
  return data;
}

/** PUT /bookings/config/platform — full-replace upsert of the platform default
 * row. Returns the persisted row (always non-null: upsert creates if absent).
 * super_admin only. */
export async function updatePlatformBookingConfig(
  payload: BookingConfigUpsert,
): Promise<BookingConfig> {
  const { data } = await api.put<BookingConfig>(
    "/bookings/config/platform",
    payload,
  );
  return data;
}

/** GET /bookings/config/tenant/{id} — one store's override row, or null when
 * that store hasn't customized (the grid would then fall back to platform).
 * Backend enforces own-tenant access for store roles; platform roles may read
 * any tenant. */
export async function fetchTenantBookingConfig(
  tenantId: string,
): Promise<BookingConfig | null> {
  const { data } = await api.get<BookingConfig | null>(
    `/bookings/config/tenant/${tenantId}`,
  );
  return data;
}

/** PUT /bookings/config/tenant/{id} — full-replace upsert of one store's
 * override. Backend enforces own-tenant write for store roles
 * (settings:update perm); platform roles may write any tenant. */
export async function updateTenantBookingConfig(
  tenantId: string,
  payload: BookingConfigUpsert,
): Promise<BookingConfig> {
  const { data } = await api.put<BookingConfig>(
    `/bookings/config/tenant/${tenantId}`,
    payload,
  );
  return data;
}

/** GET /bookings/config/effective — the merged config the grid renders off.
 * Anti-forgery contract: ``tenantId`` is REQUIRED for platform roles (they
 * must name the store they're viewing) and MUST be omitted by store roles
 * (the backend resolves their own tenant from the token). Callers therefore
 * pass ``undefined`` on the store path and the resolved target id on the HQ
 * path — the backend 403s on a mismatch. */
export async function fetchEffectiveBookingConfig(
  tenantId?: string,
): Promise<BookingConfigEffective> {
  const { data } = await api.get<BookingConfigEffective>(
    "/bookings/config/effective",
    { params: tenantId ? { tenant_id: tenantId } : undefined },
  );
  return data;
}

